ui/tempo-plugin/src/plugins/tempo-trace-query/get-trace-data.ts

```diff
@@ -0,0 +1,95 @@
+import { getUnixTime } from 'date-fns';
+import { TempoTraceQuerySpec } from '../../model/trace-query-model';
+import { TEMPO_DATASOURCE_KIND, TempoDatasourceSelector } from '../../model/tempo-selectors';
+import { TempoClient } from '../../model/tempo-client';
+
+export function getUnixTimeRange(timeRange: AbsoluteTimeRange) {
+  const { start, end } = timeRange;
+  return {
+    start: Math.ceil(getUnixTime(start)),
+    end: Math.ceil(getUnixTime(end)),
+  };
+}
+
+export const getTraceData: TraceQueryPlugin<TempoTraceQuerySpec>['getTraceData'] = async (spec, context) => {
+  if (spec.query === undefined || spec.query === null || spec.query === '') {
+    // Do not make a request to the backend, instead return an empty TraceData
+    console.error('TempoTraceQuery is undefined, null, or an empty string.');
+    return { traces: [] };
+  }
+
+  const defaultTempoDatasource: TempoDatasourceSelector = {
+    kind: TEMPO_DATASOURCE_KIND,
+  };
+
+  const client: TempoClient = await context.datasourceStore.getDatasourceClient(
+    spec.datasource ?? defaultTempoDatasource
+  );
+
+  const datasourceUrl = client?.options?.datasourceUrl;
+  if (datasourceUrl === undefined || datasourceUrl === null || datasourceUrl === '') {
+    console.error('TempoDatasource is undefined, null, or an empty string.');
+    return { traces: [] };
+  }
+
+  const getQuery = () => {
+    // if time range not defined -- only return the query from the spec
+    if (context.absoluteTimeRange === undefined) {
+      return spec.query;
+    }
+    // if the query already contains a time range (i.e.start and end times)
+    if (spec.query.includes('start=') || spec.query.includes('end=')) {
+      return spec.query;
+    }
+    // handle time range selection from UI drop down (e.g. last 5 minutes, last 1 hour )
+    const { start, end } = getUnixTimeRange(context?.absoluteTimeRange);
+    const queryStartTime = '&start=' + start;
+    const queryEndTime = '&end=' + end;
+    const queryWithTimeRange = encodeURI(spec.query) + queryStartTime + queryEndTime;
+    return queryWithTimeRange;
+  };
+
+  const enrichedTraceResponse = await client.getEnrichedTraceQuery(getQuery(), datasourceUrl);
+
+  const traces: TraceValue[] = enrichedTraceResponse.traces.map((traceValue) => {
+    const startTimeUnixMs = parseInt(traceValue.summary.startTimeUnixNano) * 1e-6; // convert to millisecond for eChart time format
+    const durationMs = traceValue.summary.durationMs;
+    const spanCount = traceValue.spanCount;
+    const errorCount = traceValue.errorCount;
+    const traceId = traceValue.summary.traceID;
+    const name = `rootServiceName="${traceValue.summary.rootServiceName.trim()}", rootTraceName="${traceValue.summary.rootServiceName.trim()}"`;
+
+    return {
+      startTimeUnixMs,
+      durationMs,
+      spanCount,
+      errorCount,
+      traceId,
+      name,
+    };
+  });
+
+  const traceData: TraceData = {
+    traces,
+    metadata: {
```

Comment:
To my other comment about ‘greater than last 7 days’ not being supported by Tempo logic, we could potentially use `metadata.notices` to convey that to the panel—didn’t finish that implementation yet but the idea there is notices can display in the panel header (info, warning, and error tags with a tooltip to show the full message) 

Doesn’t need to be in this PR, I might write up some thoughts in a separate issue, it’d be useful for TimeSeriesQuery too, for instance if one query throws errors but the others succeed, you could use a notice to not block rendering but still show the error in panel header. Or for performance if we wanted to truncate queries when too many series are returned and use an info tag in panel header so user is aware

Reply:
Interesting! I see there's a beginning iteration of this in `get-time-series-data.ts.` I'll leave the `metadata.notices` param out of this PR for now and wait for the separate issue.